### PR TITLE
Basic Implementation of permission manager

### LINKED ISSUE #3797: Implement flask-rest-jsonapi permission manager for user-related permissions
Implement permission manager  http://flask-rest-jsonapi.readthedocs.io/en/latest/permission.html for user-related permissions

### PR DESCRIPTION
**Pull request by poush**
Please read the description  before reviewing:

## **What is permission manager?**
Simply checking for the permission for each method. Instead of making decorators work on different views. The use of permission manager is based on the concept that "First prepare the things that permission decorator require to test the validation and then pass it to permission decorator"

### **How it is implemented?**

1. Separating the ```api``` intialization  to `app\api\bootstrap.py` so that it can be used in routes files  (like events.py, users.py, etc)
2. In this basic implementation, please see `tickets.py`.
3. As an example use, a decorator added to tickets.py as`api.has_permission('is_organizer', fetch='event_id', fetch_as="event_id", model=Ticket),`
4. Created a permission manager `app\api\helpers\permission_manager.py` which will do the whole work.

### **How it works?**
Let's look into `is_organizer` decorator. We know that it requires `event_id` to check if the user is associated as the organizer of that event or not.

Now if we call `/tickets/<int:id>` We do not have ```event_id``` which is required by `is_organizer`, Here we are using 
`api.has_permission('is_organizer', fetch='event_id', fetch_as="event_id", model=Ticket) `

_first arg_ => Name of the permission to use
_fetch_as_ => Name of key need to send to the decorator
_fetch_ => Name of the column in the model where we will find the value of the fetch_as key name
_model_ => Model to use to fetch the key value

There are two optional parameters
_methods_ => Comma separated string telling which methods are applicable for this check (default = GET,POST,DELETE,PATCH)

`fetch_key_url` (default=`id`) => This is the name of URL parameter whose value will be used to find the record from database.

`fetch_key_model` (default='id') => This is the name of coulmn in the database model with which the _fetch_key_url_ will be matched.

### **Before testing it**
**Please update your packages to get latest changes in flask-rest-jsonapi module**

That's all :D
Fixes #3797

### CODE REVIEW

**Comment by shubham-padia** [comment]
@poush This looks great :) .

- From the info that you mentioned, can you put up a doc about it at `docs/general/permissions.md` in this PR with a little bit of formatting, feel free to include any other info you might want to share
- In the doc, add a link to permission manager documentation of flask-rest-jsonapi docs.
- Also please make a PR to the framework regarding the changes you mentioned.

And please drop in a comment here when you want any of your co-devs to review ( this is stiil WIP I think from the do-not-merge label )

**Comment by poush** [comment]
@shubham-padia made PR there on your fork on jsonapi module. 
And Yes, everyone please review the logic and implementation. 
Added do-not-merge because I'm testing with one more scenario to make there are no any issues.

File: app/api/__init__.py [hunk 1]
```diff
@@ -48,7 +44,7 @@
 api.route(PageDetail, 'page_detail', '/pages/<int:id>')
 
 # tickets
-api.route(AllTicketList, 'all_ticket_list', '/tickets', '/events/<int:id>/tickets', '/events/<identifier>/tickets')
+api.route(AllTicketList, 'all_ticket_list', '/tickets', '/events/<int:event_id>/tickets', '/events/<identifier>/tickets')
```
Comment (shubham-padia) [inline app/api/__init__.py]: Make relevant changes in tickets api for this change

File: app/api/__init__.py [hunk 2]
```diff
@@ -173,3 +169,6 @@
 api.route(DiscountCodeDetail, 'discount_code_detail', '/discount-codes/<int:id>')
 api.route(DiscountCodeRelationship, 'discount_code_event',
           '/discount-codes/<int:id>/relationships/event')
+
+
+
```
Comment (shubham-padia) [inline app/api/__init__.py]: extra lines :)

File: app/api/helpers/permission_manager.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                return NotFoundError({'source': ''}, 'Object not found.').respond()
+
+            kwargs[fetch_as] = getattr(data, fetch)
+        else:
+            kwargs[fetch_as] = view_kwargs[kwargs['fetch_as']]
+
+    if args[0] in permissions: 
+        return permissions[args[0]](view, view_args, view_kwargs, *args, **kwargs)
+    else:
+        return ForbiddenError({'source': ''}, 'Access forbidden').respond()
```
Comment (shubham-padia) [inline app/api/helpers/permission_manager.py]: PEP8 requires new line at the end of the file

File: docs/general/permissions.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+_fetch_as_ => Name of key need to send to the decorator
+_fetch_ => Name of the column in the model where we will find the value of the fetch_as key name
+_model_ => Model to use to fetch the key value
+
+There are two optional parameters
+_methods_ => Comma separated string telling which methods are applicable for this check (default = GET,POST,DELETE,PATCH)
+
+_fetch_key_ (default=`id`) => This is identifier used to get the row containing fetch's value. This will be matched with view_kwargs to get the value and filter.
+
+### **Before testing it**
```
Comment (shubham-padia) [inline docs/general/permissions.md]: This part is not required in the docs

**Comment by niranjan94** [comment]
@poush what's the status on this. ? This issue and the other issue that is dependent on this (#3802) were assigned for the last milestone.

**Comment by poush** [comment]
@niranjan94 Updated `fetch_key` so that it can be more useful in some scenarios (see last lines of the first comment ). Updated events API to use permission manager. Please test it.

**Comment by shubham-padia** [comment]
@poush Is this ready or is something remaining, the current logic looks good to me.

**Pull request merged**